### PR TITLE
Added initial FSDP readme

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #209

### CODE REVIEW

File: docs/fsdp.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  mesh: Optional[DeviceMesh] = None,
+  reshard_after_forward: Union[bool, int] = True,
+  mp_policy: MixedPrecisionPolicy = MixedPrecisionPolicy(),
+  offload_policy: OffloadPolicy = OffloadPolicy(),
+) -> nn.Module:  # return value only used by `contract` for checks
+```
+- `fully_shard(module)` is similar to `FullyShardedDataParallel(module)`, constructing one communication bucket of `module.parameters()` (except those already assigned to a nested FSDP call).
+    - `fully_shard(module)` adds an `FSDPState` object on `module`, accessible via `fully_shard.state(module)`, instead of being an `nn.Module` wrapper. This is done via the `@contract` decorator.
+- FSDP1's `sharding_strategy` and `process_group`/`device_mesh` maps to FSDP2's `mesh` and `reshard_after_forward`.
+  - `mesh` should be 1D for FSDP and 2D for HSDP. For HSDP, we assume sharding on the 0th mesh dim and replication on the 1st mesh dim. If `mesh is None`, then FSDP2 initializes a 1D global mesh over the default process group.
```
Comment (gnadathur) [inline docs/fsdp.md:54]: @awgu ,
Are we introducing HSDP as a part of torchtrain soft launch ? If so, should add HSDP results to the PR https://github.com/pytorch/torchtrain/pull/165 ?

File: docs/fsdp.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  - `reshard_after_forward=True` means parameters are resharded (freed) after forward and re-all-gathered in backward. This with 1D mesh equals `ShardingStrategy.FULL_SHARD`. This with 2D mesh equals `ShardingStrategy.HYBRID_SHARD`.
```
Comment (gnadathur) [inline docs/fsdp.md]: Should we add the motivation for `reshard_after_forward` option ? When it makes sense to enable / disable ?

File: docs/fsdp.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+def fully_shard(
+  module: nn.Module,
+  *,
+  mesh: Optional[DeviceMesh] = None,
+  reshard_after_forward: Union[bool, int] = True,
+  mp_policy: MixedPrecisionPolicy = MixedPrecisionPolicy(),
+  offload_policy: OffloadPolicy = OffloadPolicy(),
+) -> nn.Module:  # return value only used by `contract` for checks
+```
+- `fully_shard(module)` is similar to `FullyShardedDataParallel(module)`, constructing one communication bucket of `module.parameters()` (except those already assigned to a nested FSDP call).
```
Comment (gnadathur) [inline docs/fsdp.md]: From a reader perspective, does it make sense to organize this option section as a table.
Column 1: FSDP1 option name, Column 2: FSDP2 status. Here it can be `no change` or  introduced by new flag or "removed". Current format seems hard to digest in a comparative manner.

File: docs/fsdp.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  - FSDP2 does not support `device_id`. It always moves managed modules' parameters/buffers to the current CUDA device if given a CUDA `mesh`.
+  - FSDP2 removes `limit_all_gathers`, instead using an improved memory management system without any CPU rate limiting.
+  - FSDP2 removes `use_orig_params` since the original parameters are always exposed under dim-0 per-parameter sharding.
+- Calling `model.named_parameters()` for a `model` with FSDP2 applied returns unchanged parameter names and `DTensor` sharded parameters. This means that the optimizer and gradient norm clipping see `DTensor`s.
+- `fully_shard(module)` performs a dynamic class swap on `module`. E.g., if `type(module) is Transformer`, then FSDP2 constructs a new class `FSDPTransformer` that inherits from a class `FSDP` and `Transformer` and sets `module.__class__` to be `FSDPTransformer`. This allows us to add new methods and override methods via `FSDP` without constructing an `nn.Module` wrapper.
+- Calling `model.state_dict()` for an FSDP2 `model` returns a "sharded state dict" (without any communication), whereas for FSDP1, it returns a "full state dict".
+
```
Comment (gnadathur) [inline docs/fsdp.md:84]: Are there some functionality that we dont support in FSDP2 that we want to highlight ?

File: docs/fsdp.md (hunk 1 above)
Comment (awgu) [reply docs/fsdp.md:54]: I do not think we are introducing HSDP as part of the torchtrain soft launch since it is not integrated yet. cc: @wanchaol @fegin

File: docs/fsdp.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+With these limitations in mind, we designed and implemented an FSDP rewrite removing the `FlatParameter`.  We refer to this rewrite as FSDP2 and the original as FSDP1. FSDP2 targets the same use cases as FSDP1 plus more. FSDP2 still strives for the same performance in eager mode, using several of the same techniques.
+
+Compared to FSDP1:
+- FSDP2 represents sharded parameters as `DTensor`s sharded on dim-0, allowing for easy manipulation of individual parameters, communication-free sharded state dicts, and a simpler meta-device initialization flow.
+- FSDP2 implements an improved memory management system that achieves lower and deterministic GPU memory by avoiding `recordStream` and does so without any CPU synchronization.
```
Comment (lessw2020) [inline docs/fsdp.md:10]: imo this is where you could add the visual from your tensorboard run showing the divergent gpu mem% vs the flat line of FSDP2.  
this with explanation:
Screenshot 2024-04-09 at 2 50 53 PM

File: docs/fsdp.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+- FSDP1's `sharding_strategy` and `process_group`/`device_mesh` maps to FSDP2's `mesh` and `reshard_after_forward`.
+  - `mesh` should be 1D for FSDP and 2D for HSDP. For HSDP, we assume sharding on the 0th mesh dim and replication on the 1st mesh dim. If `mesh is None`, then FSDP2 initializes a 1D global mesh over the default process group.
+  - `reshard_after_forward=True` means parameters are resharded (freed) after forward and re-all-gathered in backward. This with 1D mesh equals `ShardingStrategy.FULL_SHARD`. This with 2D mesh equals `ShardingStrategy.HYBRID_SHARD`.
+  - `reshard_after_forward=False` means parameters are not resharded (freed) after forward, not needing any all-gather in backward. This with 1D mesh equals `ShardingStrategy.SHARD_GRAD_OP`. This with 2D mesh equals `ShardingStrategy._HYBRID_SHARD_ZERO2`.
+  - (Experimental) `reshard_after_forward: int` means that parameters are resharded to a smaller world size after forward (e.g. `reshard_after_forward=8` can mean intra-node) so that the backward all-gather is over a smaller world size.
```
Comment (lessw2020) [inline docs/fsdp.md:56]: maybe point out that this is the same as hsz (heirarchical sharding) from the z++ paper?

File: docs/fsdp.md (hunk 4 above)
Comment (lessw2020) [inline docs/fsdp.md:84]: Perhaps break this section which is a long listing up into sections and group accordingly for better readability?
i.e. 
FSDP2 *removes*:
FSDP2 *adds*:
FSDP2 *changes*:

File: docs/fsdp.md (hunk 2 above)
Comment (awgu) [reply docs/fsdp.md]: Added a sentence

File: docs/fsdp.md (hunk 3 above)
Comment (awgu) [reply docs/fsdp.md]: Added a table

File: docs/fsdp.md (hunk 6 above)
Comment (awgu) [reply docs/fsdp.md:56]: Added a table to compare across FSDP1/FSDP2/DeepSpeed

Comment (lessw2020) [reply docs/fsdp.md:56]: this looks great!

File: docs/fsdp.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
+# FSDP1 -> FSDP2
+
+## Why FSDP2?
+PyTorch's fully sharded data parallelism (FSDP) API, [`FullyShardedDataParallel`](https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/fully_sharded_data_parallel.py), looks to offer a performant eager-mode implementation, including communication bucketing and communication/computation overlap. It defines a `FlatParameter` by flattening and concatenating a group of parameters to represent a communication bucket. However, this `FlatParameter` complicates applying different behaviors to individual parameters within the `FlatParameter`, e.g. parameter freezing, parameter casting, etc., hurting composability, and it complicates the internal implementation, e.g. making state dict logic thousands of lines.
```
Comment (wanchaol) [inline docs/fsdp.md]: would it be better to point user to FSDP1 doc page instead of the code directly?

Comment (wanchaol) [inline docs/fsdp.md]: maybe adding the need for state_dict communication with FlatParameter 

"making state dict logic thousands of lines and requiring additional communications for state_dict save/load"

File: docs/fsdp.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+| `sync_module_states` | removed |
+| `forward_prefetch` | not yet implemented |
+| `limit_all_gathers` | removed |
+| `use_orig_params` | removed |
+
+- `fully_shard(module)` is similar to `FullyShardedDataParallel(module)`, constructing one communication bucket from `module.parameters()` except those already assigned to a nested `fully_shard`/`FullyShardedDataParallel` call.
+    - `fully_shard(module)` adds an `FSDPState` object on `module`, accessible via `fully_shard.state(module)`, instead of being an `nn.Module` wrapper. This is done via the `@contract` decorator.
+    - Calling `model.named_parameters()` for a `model` with FSDP2 applied returns unchanged parameter names and `DTensor` sharded parameters. This means that the optimizer and gradient norm clipping see `DTensor`s.
+    - `fully_shard(module)` performs a dynamic class swap on `module`. E.g., if `type(module) is Transformer`, then FSDP2 constructs a new class `FSDPTransformer` that inherits from a class `FSDP` and `Transformer` and sets `module.__class__` to be `FSDPTransformer`. This allows us to add new methods and override methods via `FSDP` without constructing an `nn.Module` wrapper.
```
Comment (wanchaol) [inline docs/fsdp.md:52]: is this necessarily a user facing change that user need to be aware? I guess it's more like a implementation detail so we can probably hide this away from user.

Comment (awgu) [reply docs/fsdp.md:52]: I mainly wanted to mention this because I have seen users inherit from `FullyShardedDataParallel` and add their own customizations/overrides. Today, we do not have a way for the user to do that with FSDP2. We should see how we want to allow users to get this same override ability.

File: docs/fsdp.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+# Call `param_init_fn` on every module
+def param_init_fn(module: nn.Module) -> None: ...
+model = FSDP(model, auto_wrap_policy=policy, param_init_fn=param_init_fn)
+```
+After with FSDP2:
+```
+with torch.device("meta"):
+    model = Transformer()
+for module in model.modules():
+if isinstance(module, TransformerBlock):
```
Comment (tianyu-l) [inline docs/fsdp.md]: fix indentation here?

File: docs/fsdp.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
+# FSDP1 -> FSDP2
+
+## Why FSDP2?
+PyTorch's fully sharded data parallelism (FSDP) API, [`FullyShardedDataParallel`](https://pytorch.org/docs/stable/fsdp.html), looks to offer a performant eager-mode implementation, including communication bucketing and communication/computation overlap. It defines a `FlatParameter` by flattening and concatenating a group of parameters to represent a communication bucket. However, this `FlatParameter` complicates applying different behaviors to individual parameters within the `FlatParameter`, e.g. parameter freezing, parameter casting, etc., hurting composability, and it complicates the internal implementation, e.g. making state dict logic thousands of lines and requiring additional communications.
+
+With these limitations in mind, we designed and implemented an FSDP rewrite removing the `FlatParameter`.  We refer to this rewrite as FSDP2 and the original as FSDP1. FSDP2 targets the same use cases as FSDP1 plus more. FSDP2 still strives for the same performance in eager mode, using several of the same techniques.
```
Comment (tianyu-l) [inline docs/fsdp.md]: > FSDP2 still strives for the same performance in eager mode
May I ask here which is the case:
1. FSDP2 strives to achieve the same performance to FSDP1, and it's hard to get better.
2. Similar to the spirit of FSDP1, FSDP2 also strives to achieve great performance in the eager mode. Moreover, it can achieve better performance than FSDP1.

I think you meant 2, but somehow it could ambiguously be interpreted as 1. So please consider rewording a bit.

File: docs/fsdp.md (hunk 9 above)
Comment (awgu) [reply docs/fsdp.md]: Good catch!

File: docs/fsdp.md (hunk 10 above)
Comment (awgu) [reply docs/fsdp.md]: Changed to:
> FSDP2 targets the same use cases as FSDP1 plus more, and FSDP2 still strives for good performance in eager mode, using several of the same techniques.

A few lines below, added:
> We have validated FSDP2 numerics and performance using torchtrain (e.g. see this [PR](https://github.com/pytorch/torchtrain/pull/165)). For example, on some Llama-7B runs on 8x H100s, FSDP2 achieves higher MFU with 7% lower peak memory than FSDP1, matching the same loss curve.
